public/app/features/scopes/ScopesApiClient.ts

```diff
@@ -0,0 +1,105 @@
+import { Scope, ScopeDashboardBinding, ScopeNode } from '@grafana/data';
+import { getBackendSrv } from '@grafana/runtime';
+
+import { getAPINamespace } from '../../api/utils';
+
+import { NodeReason, NodesMap, SelectedScope, TreeScope } from './selector/types';
+import { getEmptyScopeObject } from './utils';
+
+const apiGroup = 'scope.grafana.app';
+const apiVersion = 'v0alpha1';
+const apiNamespace = getAPINamespace();
+const apiUrl = `/apis/${apiGroup}/${apiVersion}/namespaces/${apiNamespace}`;
+
+export class ScopesApiClient {
+  private scopesCache = new Map<string, Promise<Scope>>();
+
+  async fetchScope(name: string): Promise<Scope> {
+    if (this.scopesCache.has(name)) {
+      return this.scopesCache.get(name)!;
+    }
+
+    const response = new Promise<Scope>(async (resolve) => {
+      const basicScope = getEmptyScopeObject(name);
+
+      try {
+        const serverScope = await getBackendSrv().get<Scope>(apiUrl + `/scopes/${name}`);
+
+        const scope = {
+          ...basicScope,
+          ...serverScope,
+          metadata: {
+            ...basicScope.metadata,
+            ...serverScope.metadata,
+          },
+          spec: {
+            ...basicScope.spec,
+            ...serverScope.spec,
+          },
+        };
+
+        resolve(scope);
+      } catch (err) {
+        this.scopesCache.delete(name);
+
+        resolve(basicScope);
+      }
+    });
+
+    this.scopesCache.set(name, response);
+
+    return response;
+  }
+
+  async fetchMultipleScopes(treeScopes: TreeScope[]): Promise<SelectedScope[]> {
+    const scopes = await Promise.all(treeScopes.map(({ scopeName }) => this.fetchScope(scopeName)));
+
+    return scopes.reduce<SelectedScope[]>((acc, scope, idx) => {
+      acc.push({
+        scope,
+        path: treeScopes[idx].path,
+      });
+
+      return acc;
+    }, []);
+  }
+
+  async fetchNode(parent: string, query: string): Promise<NodesMap> {
+    try {
+      const nodes =
+        (await getBackendSrv().get<{ items: ScopeNode[] }>(apiUrl + `/find/scope_node_children`, { parent, query }))
+          ?.items ?? [];
+
+      return nodes.reduce<NodesMap>((acc, { metadata: { name }, spec }) => {
```

Comment:
Why use a reducer as opposed to create a `Map`+ a loop? IMO that is a bit more readable.

Reply:
Not sure here. Like using `Map` would change the return type and would need changes in other places (considering this will go away not sure I want to add too much work) and using `map()` then `new Map()` then converting `Map` to `Object` does not seem like a better way to do this.